include/nlohmann/detail/output/binary_writer.hpp

```diff
@@ -676,26 +676,231 @@ class binary_writer
+        {
+            oa->write_character(static_cast<CharType>(0x12)); // int64
+            oa->write_characters(
+                reinterpret_cast<const CharType*>(name.c_str()),
+                name.size() + 1u);
+
+            write_number<std::int64_t, true>(static_cast<std::int64_t>(j.m_value.number_integer));
+
+            return /*id*/ 1ul + name.size() + 1ul + sizeof(std::int64_t);
+        }
+    }
+
+    std::size_t write_bson_unsigned(const typename BasicJsonType::string_t& name, const BasicJsonType& j)
+    {
+        auto n = j.m_value.number_integer;
+        if (n <= static_cast<uint64_t>((std::numeric_limits<uint32_t>::max)()))
+        {
+            oa->write_character(static_cast<CharType>(0x10)); // int32
+            oa->write_characters(
+                reinterpret_cast<const CharType*>(name.c_str()),
+                name.size() + 1u);
+
+            write_number<std::int32_t, true>(static_cast<std::int32_t>(n));
+
+            return /*id*/ 1ul + name.size() + 1ul + sizeof(std::int32_t);
+        }
+        else
+        {
+            oa->write_character(static_cast<CharType>(0x12)); // int64
+            oa->write_characters(
+                reinterpret_cast<const CharType*>(name.c_str()),
+                name.size() + 1u);
+
+            write_number<std::int64_t, true>(static_cast<std::int64_t>(j.m_value.number_integer));
+
+            return /*id*/ 1ul + name.size() + 1ul + sizeof(std::int64_t);
+        }
+    }
+
+    std::size_t write_bson_object_internal(const typename BasicJsonType::string_t& name, const BasicJsonType& j)
+    {
+        oa->write_character(static_cast<CharType>(0x03)); // object
+        oa->write_characters(
+            reinterpret_cast<const CharType*>(name.c_str()),
+            name.size() + 1u);
+
+        auto const embedded_document_size = write_bson_object(j);
+
+        return /*id*/ 1ul + name.size() + 1ul + embedded_document_size;
+    }
+
+    std::size_t write_bson_array(const typename BasicJsonType::string_t& name, const BasicJsonType& j)
+    {
+        oa->write_character(static_cast<CharType>(0x04)); // object
+        oa->write_characters(
+            reinterpret_cast<const CharType*>(name.c_str()),
+            name.size() + 1u);
+
+
+        auto document_size_offset = oa->reserve_characters(4ul);
+        std::int32_t embedded_document_size = 5ul;
+
+        for (const auto& el : *j.m_value.array)
+        {
+            embedded_document_size += write_bson_object_entry("", el);
+        }
+
+        oa->write_character(static_cast<CharType>(0x00));
+        write_number_at<std::int32_t, true>(document_size_offset, embedded_document_size);
+
+        return /*id*/ 1ul + name.size() + 1ul + embedded_document_size;
+    }
+
+    std::size_t write_bson_object_entry(const typename BasicJsonType::string_t& name, const BasicJsonType& j)
+    {
+        switch (j.type())
+        {
+            // LCOV_EXCL_START
+            default:
+                assert(false);
```

Comment:
In case of a discarded value, we should not raise an assertion, but either do nothing or throw an exception.

Reply:
good point! I adjusted the code for this case and added a test case to check that `discarded` values are actually discarded.